tephra-core/src/main/java/co/cask/tephra/visibility/FenceWait.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.tephra.visibility;
+
+import co.cask.tephra.TransactionFailureException;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Used by a writer to wait on a fence so that changes are visible to all readers with in-progress transactions.
+ */
+public interface FenceWait {
+  /**
+   * Waits until the fence is complete, or till the timeout specified. The fence wait transaction will get re-tried
+   * several times until the timeout.
+   * <p>
+   *
+   * If a fence wait times out then it means there are still some readers with in-progress transactions that have not
+   * seen the change. In this case the wait will have to be retried using the same FenceWait object.
+   *
+   * @param timeout Maximum time to wait
+   * @param sleepTime Interval to sleep between retries
+   * @param timeUnit {@link TimeUnit} for timeout and sleepTime
+   * @throws TransactionFailureException when not able to start fence wait transaction
+   * @throws InterruptedException on any interrupt
+   * @throws TimeoutException when timeout is reached
+   */
+  void await(long timeout, long sleepTime, TimeUnit timeUnit)
```

Comment:
Better not needing to pass in the sleep time explicitly. I think it can simply be something like

```java
// Basically, Sleep time == timeout / 10, but no shorter than 1 microsecond and no longer than 500 milliseconds
long sleepTime = Maths.min(Maths.max(timeUnit.toNanos(timeout) / 10, 1000L), 500000000L);
```